core/src/main/java/org/elasticsearch/rest/RestController.java

```diff
@@ -159,29 +168,32 @@ public RestFilterChain filterChain(RestFilter executionFilter) {
     }
 
     public void dispatchRequest(final RestRequest request, final RestChannel channel, ThreadContext threadContext) {
-        if (!checkRequestParameters(request, channel)) {
-            return;
-        }
-        try (ThreadContext.StoredContext t = threadContext.stashContext()){
-            for (String key : relevantHeaders) {
-                String httpHeader = request.header(key);
-                if (httpHeader != null) {
-                    threadContext.putHeader(key, httpHeader);
-                }
+        int contentLength = request.hasContent() ? request.content().length() : 0;
+        try (Reservation reservation = inflightRequestLimiter.addBytesChecked(contentLength)) {
```

Comment:
You're right. Upon inspection of the code it looked like all code paths would produce a synchronous response. I'll push a fix soon that returns the reservation in `NettyHttpChannel` where the actual response is sent.